server/src/test/java/org/elasticsearch/indices/IndexingMemoryControllerIT.java

```diff
@@ -0,0 +1,163 @@
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.emptySet;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertNoFailures;
+
+public class IndexingMemoryControllerIT extends ESSingleNodeTestCase {
+
+    // #10312
+    public void testDeletesAloneCanTriggerRefresh() throws Exception {
+        createIndex("index",
+                    Settings.builder().put("index.number_of_shards", 1)
+                                      .put("index.number_of_replicas", 0)
+                                      .put("index.refresh_interval", -1)
+                                      .build());
+        ensureGreen();
+
+        IndicesService indicesService = getInstanceFromNode(IndicesService.class);
+        IndexService indexService = indicesService.indexService(resolveIndex("index"));
+        IndexShard shard = indexService.getShardOrNull(0);
+        assertNotNull(shard);
+
+        for (int i = 0; i < 100; i++) {
+            String id = Integer.toString(i);
+            client().prepareIndex("index").setId(id).setSource("field", "value").get();
+        }
+
+        // Force merge so we know all merges are done before we start deleting:
+        ForceMergeResponse r = client().admin().indices().prepareForceMerge().setMaxNumSegments(1).execute().actionGet();
+        assertNoFailures(r);
+
+        // Make a shell of an IMC to check up on indexing buffer usage:
+        Settings settings = Settings.builder().put("indices.memory.index_buffer_size", "1kb").build();
+
+        // TODO: would be cleaner if I could pass this 1kb setting to the single node this test created....
+        IndexingMemoryController imc = new IndexingMemoryController(settings, null, null) {
+            @Override
+            protected List<IndexShard> availableShards() {
+                return Collections.singletonList(shard);
+            }
+
+            @Override
+            protected long getIndexBufferRAMBytesUsed(IndexShard shard) {
+                return shard.getIndexBufferRAMBytesUsed();
+            }
+
+            @Override
+            protected void writeIndexingBufferAsync(IndexShard shard) {
+                // just do it sync'd for this test
+                shard.writeIndexingBuffer();
+            }
+
+            @Override
+            protected Cancellable scheduleTask(ThreadPool threadPool) {
+                return null;
+            }
+        };
+
+        for (int i = 0; i < 100; i++) {
+            String id = Integer.toString(i);
+            client().prepareDelete("index", id).get();
+        }
+
+        final long indexingBufferBytes1 = shard.getIndexBufferRAMBytesUsed();
+
+        imc.forceCheck();
```

Comment:
I think there is a (very small) risk that the `IndexingMemoryController` created by the node was triggered on one  of the deletes and that this causes the writeIndexingBuffer/refresh call triggered here to return immediately. It does take a lot of bad circumstances though (of which the heap size alone will likely prevent this), but I still think we should address it. Maybe the TODO above is doable if this test is moved to its own class? Or we could simply add a node-setting that essentially disables it by setting the limit very high?

Reply:
Good point. I pushed https://github.com/elastic/elasticsearch/pull/50769/commits/4d04e97c214626fd184de151d6fe36949e0267cb.